nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/RecordStats.java

```diff
@@ -0,0 +1,176 @@
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (input == null) {
+            return;
+        }
+
+        try {
+            Map<String, RecordPath> paths = getRecordPaths(context);
+            Map<String, String> stats = getStats(input, paths, context, session);
+
+            input = session.putAllAttributes(input, stats);
+
+            session.transfer(input, REL_SUCCESS);
+
+        } catch (Exception ex) {
+            getLogger().error("Error processing stats.", ex);
+            session.transfer(input, REL_FAILURE);
+        }
+
+    }
+
+    protected Map<String, RecordPath> getRecordPaths(ProcessContext context) {
+        return context.getProperties().keySet()
+            .stream().filter(p -> p.isDynamic() && !p.getName().contains(RECORD_READER.getName()))
+            .collect(Collectors.toMap(
+                e -> e.getName(),
+                e ->  {
+                    String val = context.getProperty(e).getValue();
+                    return cache.getCompiled(val);
+                })
+            );
+    }
+
+    protected Map<String, String> getStats(FlowFile flowFile, Map<String, RecordPath> paths, ProcessContext context, ProcessSession session) {
+        try (InputStream is = session.read(flowFile)) {
+            RecordReaderFactory factory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+            RecordReader reader = factory.createRecordReader(flowFile, is, getLogger());
+
+            Map<String, Integer> retVal = new HashMap<>();
+            Record record;
+
+            int recordCount = 0;
+            while ((record = reader.nextRecord()) != null) {
+                for (Map.Entry<String, RecordPath> entry : paths.entrySet()) {
+                    RecordPathResult result = entry.getValue().evaluate(record);
+                    Optional<FieldValue> value = result.getSelectedFields().findFirst();
+                    if (value.isPresent() && value.get().getValue() != null) {
+                        String approxValue = value.get().getValue().toString();
+                        String key = String.format("%s.%s", entry.getKey(), approxValue);
+                        Integer stat = retVal.containsKey(key) ? retVal.get(key) : 0;
+                        Integer baseStat = retVal.containsKey(entry.getKey()) ? retVal.get(entry.getKey()) : 0;
+                        stat++;
+                        baseStat++;
+
+                        retVal.put(key, stat);
+                        retVal.put(entry.getKey(), baseStat);
+                    }
+                }
+
+                recordCount++;
+            }
+            retVal.put(RECORD_COUNT_ATTR, recordCount);
+
+            return retVal.entrySet().stream()
+                .collect(Collectors.toMap(
+                    e -> e.getKey(),
+                    e -> e.getValue().toString()
```

Comment:
For the non-number values, such as the 'sport' dataset, the list can be huge, order of thousands or more. That can be too big to store as FlowFile attributes. At lease we need a fixed number of attributes for this processor can add. It would be more helpful if we can limit the N number of values (i.e. soccer, football, basketball ... etc) and report the highest N variables, to report better stats with a long-tail dataset.

Reply:
Ok. We can add a limit there.